crates/starknet-types-core/src/felt/primitive_conversions.rs

```diff
@@ -0,0 +1,334 @@
+            res = -res;
+        }
+        res
+    }
+}
+
+// TryFrom<Felt> for primitive
+
+#[derive(Debug, Copy, Clone)]
+pub struct PrimitiveFromFeltError;
+
+#[cfg(feature = "std")]
+impl std::error::Error for PrimitiveFromFeltError {}
+
+impl core::fmt::Display for PrimitiveFromFeltError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "Failed to convert `Felt` into primitive type")
+    }
+}
+
+macro_rules! impl_try_felt_into_unsigned {
+    ($into: ty) => {
+        impl TryFrom<Felt> for $into {
+            type Error = PrimitiveFromFeltError;
+
+            fn try_from(value: Felt) -> Result<$into, Self::Error> {
+                let bytes_be = value.to_bytes_le();
+                let (bytes_to_return, bytes_to_check) =
+                    bytes_be.split_at(core::mem::size_of::<$into>());
+
+                if bytes_to_check.iter().all(|&v| v == 0) {
+                    Ok(<$into>::from_le_bytes(bytes_to_return.try_into().unwrap()))
+                } else {
+                    Err(PrimitiveFromFeltError)
+                }
+            }
+        }
+    };
+}
+
+impl_try_felt_into_unsigned!(u8);
+impl_try_felt_into_unsigned!(u16);
+impl_try_felt_into_unsigned!(u32);
+impl_try_felt_into_unsigned!(u64);
+impl_try_felt_into_unsigned!(usize);
+impl_try_felt_into_unsigned!(u128);
+
+const MINUS_TWO_BYTES_REPR: [u8; 32] = [
+    255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+    255, 255, 255, 255, 255, 16, 0, 0, 0, 0, 0, 0, 8,
+];
+
+macro_rules! impl_try_felt_into_signed {
+    ($into: ty) => {
+        impl TryFrom<Felt> for $into {
+            type Error = PrimitiveFromFeltError;
+
+            // Let's call `s` the `size_of_type`.
+            //
+            // There is 3 ways the value we are looking for can be encoded in the Felt bytes:
+            // 1. Positive numbers:
+            //  Bytes [s..] are all set to zero, and byte [s] value must not exceed 127
+            // 2. Negative numbers:
+            //  Bytes [s..] should match `MINUS_TWO_BYTES_REPR`, and byte [s] value must not be less than 128
+            // 3. -1:
+            //  Bytes are those of Felt::MAX
+            fn try_from(value: Felt) -> Result<Self, Self::Error> {
+                let size_of_type = core::mem::size_of::<$into>();
+                let bytes_le = value.to_bytes_le();
+
+                // Positive numbers
+                if bytes_le[size_of_type..].iter().all(|&v| v == 0)
+                    && bytes_le[size_of_type - 1] <= 0x7f
+                {
+                    Ok(<$into>::from_le_bytes(
+                        bytes_le[..size_of_type].try_into().unwrap(),
+                    ))
+                }
+                // Negative numbers
+                else if bytes_le[size_of_type..] == MINUS_TWO_BYTES_REPR[size_of_type..]
+                    && bytes_le[size_of_type - 1] >= 0x80
+                {
```

Comment:
```suggestion
            fn try_from(value: Felt) -> Result<Self, Self::Error> {
                // We'll split the conversion in 3 case:
                // Felt >= 0
                // Felt == -1
                // Felt < -1

                // Get the size of the type we'll convert the felt into
                let size_of_type = core::mem::size_of::<$into>();
                // Convert the felt as bytes
                let bytes_le = value.to_bytes_le();

                // Case 1: Felt >= 0
                // Our felt type can hold values up to roughly 2**252 bits which is encoded in 32 bytes.
                // The type we'll convert the value into can hold `size_of_type` (= { 1, 2, 4, 8, 16 }) bytes.
                // If all the bytes after the last byte that our target type can fit are 0 it means that the
                // number is positive.
                // The target type is a signed type which means that the leftmost bit of the last byte
                // (remember we're in little endian) is used for the sign (0 for positive numbers)
                // so the last byte can hold a value up to 0b01111111
                if bytes_le[size_of_type..].iter().all(|&v| v == 0)
                    && bytes_le[size_of_type - 1] <= 0b01111111
                {
                    // We take the `size_of_type` first bytes because they contain the useful value.
                    Ok(<$into>::from_le_bytes(
                        bytes_le[..size_of_type].try_into().unwrap(),
                    ))
                }
                // Case 2: Felt == -1
                // This is the little endian representation of Felt::MAX. And Felt::MAX is exactly -1
                // [
                //    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                //    0, 0, 0, 0, 0, 17, 0, 0, 0, 0, 0, 0, 8,
                // ]
                // The only valid Felt with those most significant bytes is Felt::MAX.
                // All other arrays with those most significant bytes are invalid Felts because they would be >= Prime.
                else if bytes_le[24..] == [17, 0, 0, 0, 0, 0, 0, 8] {
                    return Ok(-1);
                }
                // Case 3: Felt < -1
                // Similarly to how we checked that the number was positive by checking the bytes after the
                // `size_of_type` byte we check that all the bytes after correspond to the bytes of a negative felt
                // The leftmost bit is use for the sign, as it's a negative number it has to be 1.
                else if bytes_le[size_of_type..] == MINUS_TWO_BYTES_REPR[size_of_type..]
                    && bytes_le[size_of_type - 1] >= 0b10000000
                {
                    // We take the `size_of_type` first bytes because they contain the useful value.

```

Reply:
done